### PR TITLE
[SPARK-30570][BUILD] Update scalafmt plugin to 1.0.3 with onlyChangedFiles feature

### PR DESCRIPTION
**Pull request by koeninger**
### What changes were proposed in this pull request?
Update the scalafmt plugin to 1.0.3 and use its new onlyChangedFiles feature rather than --diff

### Why are the changes needed?
Older versions of the plugin either didn't work with scala 2.13, or got rid of the --diff argument and didn't allow for formatting only changed files

### Does this PR introduce any user-facing change?
The /dev/scalafmt script no longer passes through arbitrary args, instead using the arg to select scala version.  The issue here is the plugin name literally contains the scala version, and doesn't appear to have a shorter way to refer to it.   If @srowen or someone else with better maven-fu has an idea I'm all ears.

### How was this patch tested?
Manually, e.g. edited a file and ran

dev/scalafmt

or

dev/scalafmt 2.13

### CODE REVIEW

**Review by srowen — APPROVED** [review]
Looks good pending tests.

**Comment by dongjoon-hyun** [comment]
Hi, @koeninger and @srowen . This PR looks good, but is blocked by the following PR because GitHub Action is failing on this PR.
- https://github.com/apache/spark/pull/27281

Let's wait and see the GitHub Job successful run after we merge https://github.com/apache/spark/pull/27281 first.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
@dongjoon-hyun, `dev/scalafmt` is not being tested in Jenkins PR builder or Github Action I believe. It's just an optional dev script.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Ah, I got what you mean:


```
[ERROR] Plugin org.antipathy:mvn-scalafmt_2.12:1.0.3 or one of its dependencies could not be resolved: Could not find artifact org.antipathy:mvn-scalafmt_2.12:jar:1.0.3 in google-maven-central (https://maven-central.storage-download.googleapis.com/repos/central/data/) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException
##[error]Process completed with exit code 1.
```

Okay, seems good to block it due to Github Actions.

**Comment by srowen** [comment]
Jenkins, retest this please

**Comment by HyukjinKwon** [comment]
I'm periodically re-running the Github Actions. I will merge it once they pass.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Now, everything goes green after @HyukjinKwon 's patch.
Thank you, @koeninger , @srowen , @HyukjinKwon . Merged to master.

**Pull request closed without merging**
